felix/bpf/ut/pol_prog_test.go

```diff
@@ -2639,101 +2641,170 @@ func TestPolicyProgramsFlowLog(t *testing.T) {
-												Protocol: &proto.Protocol{
-													NumberOrName: &proto.Protocol_Name{Name: "udp"},
-												},
+										},
+									}},
+								},
+								{
+									Name:      "UDP allow - staged",
+									Staged:    true,
+									NoMatchID: 0x1717,
+									Rules: []polprog.Rule{{
+										MatchID: 0x17,
+										Rule: &proto.Rule{
+											Action: "Allow",
+											Protocol: &proto.Protocol{
+												NumberOrName: &proto.Protocol_Name{Name: "udp"},
 											},
-										}},
-									},
-									{
-										Name:  "allow all",
-										Rules: []polprog.Rule{{Rule: &proto.Rule{Action: "Allow"}, MatchID: 1}},
-									},
+										},
+									}},
+								},
+								{
+									Name:  "allow all",
+									Rules: []polprog.Rule{{Rule: &proto.Rule{Action: "Allow"}, MatchID: 1}},
 								},
 							},
 						},
 					},
-					allowedPackets: []testFlowLogCase{
-						{
-							packet:  udpPkt("10.0.0.1:31245", "10.0.0.2:80"),
-							matches: []uint64{1234, 0xdead, 0x600d, 0x7455, 0x66, 0x17, 1},
-						},
-						{
-							packet:  tcpPkt("10.0.0.2:80", "10.0.0.1:31245"),
-							matches: []uint64{1234, 0xdead, 0x600d, 0x7455, 0x6, 0x1717, 1},
-						},
-						{
-							packet:  icmpPkt("10.0.0.1", "10.0.0.2"),
-							matches: []uint64{1234, 0xdead, 0x600d, 0x7455, 0x66, 0x1717, 1},
-						},
+				},
+				allowedPackets: []testFlowLogCase{
+					{
+						packet:  udpPkt("10.0.0.1:31245", "10.0.0.2:80"),
+						matches: []uint64{1234, 0xdead, 0x600d, 0x7455, 0x66, 0x17, 1},
+					},
+					{
+						packet:  tcpPkt("10.0.0.2:80", "10.0.0.1:31245"),
+						matches: []uint64{1234, 0xdead, 0x600d, 0x7455, 0x6, 0x1717, 1},
+					},
+					{
+						packet:  icmpPkt("10.0.0.1", "10.0.0.2"),
+						matches: []uint64{1234, 0xdead, 0x600d, 0x7455, 0x66, 0x1717, 1},
 					},
 				},
 			},
-		*/
+		},
 	}
 
 	for _, test := range tests {
 		t.Run(fmt.Sprintf("name=%s", test.name), func(t *testing.T) { runTest(t, test.test) })
 	}
 }
 
+func TestPolicyProgramsWithStagedPolicy(t *testing.T) {
+	for i, p := range polProgramTests {
+		if p.ForIPv6 {
+			// XXX skip for now
+			continue
+		}
+		if len(p.Policy.Tiers) > 0 {
+
+			// Make a copy of the tiers
+			tiers := p.Policy.Tiers
+			p.Policy.Tiers = make([]polprog.Tier, len(tiers))
+			copy(p.Policy.Tiers, tiers)
+
+			p.Policy.Tiers[0].Policies = append(
+				[]polprog.Policy{{
+					Name: "staged allow all",
+					Rules: []polprog.Rule{{
+						Rule: &proto.Rule{
+							Action: "Allow",
+						},
+						MatchID: 0x600d,
+					}},
+					Staged:    true,
+					NoMatchID: 0x666,
+				}},
+				p.Policy.Tiers[0].Policies...)
+
+			if len(p.Policy.Tiers[0].Policies) == 1 {
+				p.Policy.Tiers[0].EndAction = polprog.TierEndPass
+			}
+			t.Run(fmt.Sprintf("Staged policy first in first tier %d:Policy=%s", i, p.PolicyName),
+				func(t *testing.T) { runTest(t, wrap(p)) },
+			)
+		}
+	}
+
+	for i, p := range polProgramTests {
+		if p.ForIPv6 {
```

Comment:
ditto

Reply:
Same thing. It's enabled and passing now.